File: Algorithms/challenges/lc003_longest_substring_without_repeating_characters.py
"""
Time: O(n)
Space: O(1)  because max of len(set) is 256, size of ASCII.

Given a string, find the length of the longest substring without repeating characters.

Examples:

Given "abcabcbb", the answer is "abc", which the length is 3.

Given "bbbbb", the answer is "b", with the length of 1.

Given "pwwkew", the answer is "wke", with the length of 3. Note that the answer must be a substring, "pwke" is a subsequence and not a substring.
"""


class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        n = len(s)
        i = -1  # critical: i is exclusive left boundary of unique set.
        m = {}  # or use a list of 256 0s.
        maxlen = 0
        for j in range(n):
            i = max(m.get(s[j], -1), i)  # critical
            m[s[j]] = j
            maxlen = max(maxlen, j-i)
        return maxlen

File: Algorithms/challenges/lc003_longest_substring_without_repeating_characters.cc
#include <string>
#include <unordered_map>
#include <vector>
#include <algorithm>

using std::string;

class Solution {
public:
  int lengthOfLongestSubstring(string s) {
      int i = -1;
      std::unordered_map<char, int> m;
      unsigned maxlen = 0;
      for (unsigned j = 0; j != s.length(); j++)
        {
          if (m.find(s[j]) != m.end())
            i = std::max(m[s[j]], i);
          m[s[j]] = j;
          maxlen = std::max(maxlen, j - i);
        }
      return maxlen;
    }
};


class Solution2 {
public:
  int lengthOfLongestSubstring(string s) {
    int i = -1;
    std::vector<int> m(256, -1);
    unsigned maxlen = 0;
    for (unsigned j = 0; j != s.length(); j++)
      {
        i = std::max(m[s[j]], i);
        m[s[j]] = j;
        maxlen = std::max(maxlen, j - i);
      }
    return maxlen;
  }
};
